contracts/governance/src/participant.rs

```diff
@@ -0,0 +1,152 @@
+use soroban_sdk::{contracttype, storage::Persistent, Address, Env, Map};
+
+use crate::{DataKey, Error};
+
+#[derive(Debug, Clone)]
+#[contracttype]
+pub struct Participant {
+    address: Address,
+    whitelisted: bool,
+    current_balance: i128,
+}
+
+impl Participant {
+    pub fn new(address: Address) -> Self {
+        Participant {
+            address,
+            whitelisted: false,
+            current_balance: 0,
+        }
+    }
+
+    pub fn is_whitelisted(&self) -> bool {
+        self.whitelisted
+    }
+
+    pub fn whitelist(&mut self) {
+        self.whitelisted = true
+    }
+
+    pub fn address(&self) -> &Address {
+        &self.address
+    }
+
+    pub fn balance(&self) -> i128 {
+        self.current_balance
+    }
+
+    pub fn increase_balance(&mut self, inc: i128) -> Result<(), Error> {
+        if inc <= 0 {
+            return Err(Error::InvalidAmount);
+        }
+        self.current_balance = self.current_balance.checked_add(inc).unwrap();
+        Ok(())
+    }
+
+    pub fn decrease_balance(&mut self, dec: i128) -> Result<(), Error> {
+        if dec <= 0 {
+            return Err(Error::InvalidAmount);
+        }
+        if dec > self.balance() {
+            return Err(Error::InsufficientFunds);
+        }
+        self.current_balance = self.current_balance.checked_sub(dec).unwrap();
+        Ok(())
+    }
+}
+
+/// Participant repository provides a pattern for
+/// condensing all storage logic for the participant.
+///
+/// It consists on adquiring the participants data structure
+/// from the Soroban storage, do all needed operations and
+/// finally persist them once the object is dropped.
+pub struct Repository<'a> {
+    storage: &'a Persistent,
+    data_key: DataKey,
+    participants_storage: Map<Address, Participant>,
+}
+
+impl<'a> Repository<'a> {
```

Comment:
Would `Storage` be a name more aligned with soroban nomenclature ? 

@geofmureithi @mariopil

Reply:
I don't have preference here, both are fine.